File: server/telegram/bot.ts
import type { Update, Message } from 'telegram-typings'
import { telegramClient } from './client'
import { execute } from './command'

const extractCommands = (
    msg: Message | undefined,
    botName: string,
): { cmd: string; arg: string }[] => {
    if (!msg || !msg.text || !msg.entities) return []
    const text = msg.text
    const commands = msg.entities
        .filter((entity) => entity.type === 'bot_command')
        .map((entity) => {
            const cmds = text.substr(entity.offset, entity.length).split('@')
            if (
                cmds.length === 1 ||
                (cmds.length === 2 && cmds[1] === botName)
            ) {
                return { cmd: cmds[0], off: entity.offset, len: entity.length }
            } else {
                return null
            }
        })
        .filter(Boolean) as { cmd: string; off: number; len: number }[]

    const cmds: { cmd: string; arg: string }[] = []
    let i = 0
    while (i < commands.length) {
        const command = commands[i]
        const nextOffset =
            i + 1 < commands.length ? commands[i + 1].off : text.length
        const arg = text.substring(command.off + command.len, nextOffset)
        cmds.push({ cmd: command.cmd, arg })
        i++
    }
    return cmds
}

const handleMsg = async (msg: Message | undefined) => {
    const cmds = extractCommands(msg, 'FeedBoxBot')
    const tasks = cmds.map((cmd) => execute(cmd.cmd, cmd.arg, msg!))
    await Promise.all(tasks)
}

export const telegramBot = {
    registerWebhook: async () => {
        await Promise.all([
            telegramClient.send('setWebhook', {
                url: `${process.env.SERVER}/webhook/telegram/${process.env.TELEGRAM_WEBHOOK_PATH}`,
            }),
            telegramClient.send('setMyCommands', {
                commands: [
                    { command: 'list', description: 'list all feeds' },
                    { command: 'add', description: '[url] subscribe urls' },
                    { command: 'del', description: '[url] unsubscribe urls' },
                    {
                        command: 'export',
                        description: 'export feed list as OPML',
                    },
                ],
            }),
        ])
    },
    handleWebhook: async (payload: Update) => {
        await Promise.all([
            handleMsg(payload.message),
            handleMsg(payload.edited_message),
            handleMsg(payload.channel_post),
            handleMsg(payload.edited_channel_post),
            // handleCallback(payload.callback_query)
        ])
    },
}

File: service/telegram/bot.go
package telegram

import (
	"os"
	"strings"
	"unicode/utf16"

	"github.com/dhcmrlchtdj/feedbox/service/monitor"
)

func RegisterWebhook() error {
	err := Client.SetWebhook(
		&SetWebhookPayload{
			URL: os.Getenv("SERVER") + "/webhook/telegram/" + os.Getenv("TELEGRAM_WEBHOOK_PATH"),
		})
	if err != nil {
		return err
	}

	err = Client.SetMyCommands(
		&SetMyCommandsPayload{
			Commands: []BotCommand{
				{"list", "list all feeds"},
				{"add", "[url] subscribe url"},
				{"remove", "[url] unsubscribe url"},
				{"remove_all", "unsubscribe all"},
				{"export", "export feed list as OPML"},
			},
		})
	if err != nil {
		return err
	}

	return nil
}

func HandleWebhook(payload *Update) {
	handleMessage(payload.Message)
	handleMessage(payload.EditedMessage)
	handleMessage(payload.ChannelPost)
	handleMessage(payload.EditedChannelPost)
}

func handleMessage(message *Message) {
	if message == nil {
		return
	}
	for _, entity := range message.Entities {
		if entity.Type == "bot_command" {
			text := utf16.Encode([]rune(message.Text))
			cmd := string(utf16.Decode(text[entity.Offset:entity.Length]))
			if i := strings.Index(cmd, "@"); i != -1 {
				name := cmd[i+1:]
				cmd = cmd[:i]
				if name != os.Getenv("TELEGRAM_BOT_NAME") {
					continue
				}
			}
			arg := string(utf16.Decode(text[entity.Offset+entity.Length:]))
			arg = strings.TrimSpace(arg)
			if err := executeCommand(cmd, arg, message); err != nil {
				monitor.Client.Error(err)
			}
			return
		}
	}
}
